app/controllers/internal/events_controller.rb

```diff
@@ -14,8 +18,75 @@ def show
       @page_title = @event.name
     end
 
+    def new
+      @event = Event.new(venue_type: Event::VENUE_TYPES[:existing])
+      @event.building = EventBuilding.new
+    end
+
+    def approve
+      api_event = GetIntoTeachingApiClient::TeachingEventsApi.new.get_teaching_event(params[:format])
+      api_event.status_id = GetIntoTeachingApiClient::Constants::EVENT_STATUS["Open"]
+      submit_to_api(api_event)
+      redirect_to internal_events_path(success: true)
+    end
+
+    def create
+      @event = Event.new(event_params)
+      @event.building = format_building(event_params)
+      if submit_pending
+        redirect_to internal_events_path(success: :pending)
+      else
+        render action: :new
+      end
+    end
+
+    def edit
+      api_event = GetIntoTeachingApiClient::TeachingEventsApi.new.get_teaching_event(params[:id])
+      @event = Event.initialize_with_api_event(api_event)
+      render :new
+    end
+
   private
 
+    def submit_pending
+      return if invalid?
+
+      @event.status_id = GetIntoTeachingApiClient::Constants::EVENT_STATUS["Pending"]
+      api_event = @event.to_api_event
+      submit_to_api(api_event)
+    end
+
+    def invalid?
+      invalid_building = @event.building.nil? ? false : @event.building.invalid?
+      (@event.invalid? || invalid_building)
+    end
+
+    def submit_to_api(body)
+      GetIntoTeachingApiClient::TeachingEventsApi.new.upsert_teaching_event(body)
+      true
+    rescue GetIntoTeachingApiClient::ApiError => e
+      @event.map_api_errors_to_attributes(e) if e.code == 400
+      false
+    end
+
+    def format_building(event_params)
```

Comment:
This method is doing a lot and its a bit hard to follow. I think it would benefit from refactoring into a couple of methods and I also think this should encapsulate this in the `Internal::Event` so that you can just do `Event.new(event_params, @buildings)` and it takes care of constructing the related building?